blog: Streamlining access to Grafana & Prometheus (#756)
mattermost/mattermost-developer-documentation: site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md

```
ENABLED_DOCKER_SERVICES = mysql postgres inbucket prometheus grafana
```

Alternatively, you can follow the instructions in [config.mk](https://github.com/mattermost/mattermost-server/blob/master/config.mk). (Note that it's not necessary to specify `minio`, `openldap` or `elasticsearch` -- these are added automatically as needed.)

The build tooling does more than just spin up two new containers, however. [Various configuration files](https://github.com/mattermost/mattermost-server/tree/master/build/docker) also automate the following:
* configuring Prometheus to scrape the Mattermost `:8067/metrics` endpoint (requires access to the enterprise source)
* enabling anonymous access to Grafana (you can still login with the default `admin`/`admin` credentials if needed)
* configuring Grafana with the containerized Prometheus as the default data source
* installing our canonical Grafana dashboards ([Mattermost Performance KPI Metrics](https://grafana.com/grafana/dashboards/2539), [Mattermost Performance Monitoring](https://grafana.com/grafana/dashboards/2542), [Mattermost Performance Monitoring (Bonus Metrics)](https://grafana.com/grafana/dashboards/2545))

> This means that simply running `make run-server` gets you immediate access to the Mattermost Performance dashboards at `http://localhost:3000`:

Reviewer: One question, do you have to enable "Metrics" in your mattermost to make it work, right? Probably adding that info here would be great, something like "Set your `MetricsSettings.Enable` to `true` in your configuration file (or through the admin console)".

Author: Excellent point! I'll also add a note here about having to restart the server when installing a license -- for some reason, metrics (even if enabled) don't start until the next server start.

Reviewer 2: From my experience it's sufficient to trigger a config save to make the endpoint work. This is assuming the server build is enterprise ready. I think we are missing some logic to start the server when a valid license is added. It might be worth consider fixing this edge case.

Author: Indeed! I'll look at that edge cases vs. permanently documenting the pain point here. Otherwise, I've pushed changes to document enabling the metrics. Thanks!

---
title: "Streamlining Developer Access to Prometheus and Grafana"
slug: streamlining-developer-access-to-prometheus-and-grafana
date: 2021-01-21T00:00:00-04:00
categories:
    - "testing"
    - "performance"
author: Jesse Hallam
github: lieut-data
community: jesse.hallam

> summary: With access to the enterprise source code, the developer build tooling now automates the setup of Prometheus and Grafana for performance monitoring. Even the canonical Grafana dashboards are setup without any manual configuration required!

Reviewer 3: ```suggestion
summary: With access to the Enterprise source code, the developer build tooling now automates the setup of Prometheus and Grafana for performance monitoring. Even the canonical Grafana dashboards are setup without any manual configuration required!
```

---

Our [Makefile](https://github.com/mattermost/mattermost-server/blob/master/Makefile) entrypoint for developing against the [Mattermost Server](https://github.com/mattermost/mattermost-server) already tries to simplify things for developers as much as possible.

For example, when invoking `make run-server`, this build tooling takes care of all of the following (among other things!):

> * downloading a suitable version of `mmctl` for API-driven command line interaction

Reviewer 3: ```suggestion
* Downloading a suitable version of `mmctl` for API-driven command line interaction.
```

> * checking your installed Go version to avoid cryptic error messages

Reviewer 3: ```suggestion
* Checking your installed Go version to avoid cryptic error messages.
```

> * downloading and starting various Docker containers:

Reviewer 3: ```suggestion
* Downloading and starting various Docker containers:
```

>     - `mysql` and `postgres`, both for a backend to the server and also to automate tests

Reviewer 3: ```suggestion
    - `mysql` and `postgres`, both for a backend to the server and also to automate tests.
```

>     - `inbucket` to simplify email testing without routing outside your machine

Reviewer 3: ```suggestion
    - `inbucket` to simplify email testing without routing outside your machine.
```

>     - `minio` to expose an S3-compatible interface to your local disk for for hosting uploaded files

Reviewer 2: ```suggestion
    - `minio` to expose an S3-compatible interface to your local disk for hosting uploaded files
```

>     - `minio` to expose an S3-compatible interface to your local disk for for hosting uploaded files

Reviewer 3: ```suggestion
    - `minio` to expose an S3-compatible interface to your local disk for hosting uploaded files.
```

>     - `openldap` to simplify testing alternate modes of authentication (requires access to the enterprise source)

Reviewer 3: ```suggestion
    - `openldap` to simplify testing alternate modes of authentication (requires access to the enterprise source).
```

>     - `elasticsearch` to expose an improved search backend for posts and channels (requires access to the enterprise source)

Reviewer 3: ```suggestion
    - `elasticsearch` to expose an improved search backend for posts and channels (requires access to the Enterprise source code).
```

> * linking your server to an automatically detected `mattermost-webapp` directory

Reviewer 3: ```suggestion
* Linking your server to an automatically-detected `mattermost-webapp` directory.
```

> * optionally starting the server in the foreground if `RUN_SERVER_IN_BACKGROUND` is `false`

Reviewer 3: ```suggestion
* Optionally starting the server in the foreground if `RUN_SERVER_IN_BACKGROUND` is `false`.
```

> * piping the structured server logs through a `logrus` decorator for easier reading

Reviewer 3: ```suggestion
* Piping the structured server logs through a `logrus` decorator for easier reading.
```

> * invoking the `go` compiler to run the server code in your local repository

Reviewer 3: ```suggestion
* Invoking the `go` compiler to run the server code in your local repository.
```

> ### Adding Prometheus & Grafana

Reviewer 3: ```suggestion
### Adding Prometheus and Grafana
```

> As of https://github.com/mattermost/mattermost-server/pull/16649, this build tooling now supports automating the setup of Prometheus and Grafana Docker containers. Together, these tools form the backbone of our [Performance Monitoring (E20)](https://docs.mattermost.com/deployment/metrics.html) functionality, scraping metrics from Mattermost to help give enterprise customers insight into the performance of their clusters. Automating this setup makes it easier for developers to test in a production-like environment, and to help extend the monitoring available to customers.

Reviewer 2: I am obviously biased but since metrics are also a requirement for running certain type of load-tests (making full use of the coordinator) it could be nice to mention that here as an added benefit of having this.

> Since the corresponding Mattermost functionality is only available with access to the enterprise source code, these containers aren't enabled by default. To turn them on, export the following environment variable into your shell:

Reviewer 3: ```suggestion
Since the corresponding Mattermost functionality is only available with access to the Enterprise source code, these containers aren't enabled by default. To turn them on, export the following environment variable into your shell:
```

```
export ENABLED_DOCKER_SERVICES="mysql postgres inbucket prometheus grafana"
```

> 

Reviewer 3: ```suggestion
```

> Alternatively, you can follow the instructions in [config.mk](https://github.com/mattermost/mattermost-server/blob/master/config.mk). (Note that it's not necessary to specify `minio`, `openldap` or `elasticsearch` -- these are added automatically when an enterprise-enabled repository is detected.)

Reviewer 3: ```suggestion
Alternatively, you can follow the instructions in [config.mk](https://github.com/mattermost/mattermost-server/blob/master/config.mk). Note that it's not necessary to specify `minio`, `openldap`, or `elasticsearch` - these are added automatically when an Enterprise-enabled repository is detected.
```

> The build tooling does more than just spin up two new containers, however. [Various configuration files](https://github.com/mattermost/mattermost-server/tree/master/build/docker) also automate the following:

Reviewer 3: ```suggestion
The build tooling does more than just spin up two new containers, however. [Various configuration files](https://github.com/mattermost/mattermost-server/tree/master/build/docker) also automate the following:

```

> * configuring Prometheus to scrape the Mattermost `:8067/metrics` endpoint (requires access to the enterprise source)

Reviewer 3: ```suggestion
* Configuring Prometheus to scrape the Mattermost `:8067/metrics` endpoint (requires access to the Enterprise source code).
```

> * enabling anonymous access to Grafana (you can still login with the default `admin`/`admin` credentials if needed)

Reviewer 3: ```suggestion
* Enabling anonymous access to Grafana (you can still log in with the default `admin`/`admin` credentials if needed).
```

> * configuring Grafana with the containerized Prometheus as the default data source

Reviewer 3: ```suggestion
* Configuring Grafana with the containerized Prometheus as the default data source.
```

> * installing our canonical Grafana dashboards ([Mattermost Performance KPI Metrics](https://grafana.com/grafana/dashboards/2539), [Mattermost Performance Monitoring](https://grafana.com/grafana/dashboards/2542), [Mattermost Performance Monitoring (Bonus Metrics)](https://grafana.com/grafana/dashboards/2545))

Reviewer 3: ```suggestion
* Installing our canonical Grafana dashboards ([Mattermost Performance KPI Metrics](https://grafana.com/grafana/dashboards/2539), [Mattermost Performance Monitoring](https://grafana.com/grafana/dashboards/2542), [Mattermost Performance Monitoring (Bonus Metrics)](https://grafana.com/grafana/dashboards/2545)).
```

This means that simply running `make run-server` gets you immediate access to the Mattermost Performance dashboards at `http://localhost:3000`:

![Grafana home dashboard](/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana/grafana.png)

Prometheus is also available at `http://localhost:9090`:

![Prometheus landing page](/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana/prometheus.png)

### Enabling Metrics

> Although the tooling is all setup to scrape metrics and display the dashboards, you'll need to do some one-time work to enable metrics within the Mattermost server.

Reviewer 3: ```suggestion
Although the tooling is all set up to scrape metrics and display the dashboards, you'll need to do some one-time work to enable metrics within the Mattermost server.
```

> First, be sure you have an Enterprise license installed. Staff members should have access to the shared developer license, but you can also request a trial license in-product. Browse to `/admin_console/about/license` on your local Mattermost instance to setup the license.

Reviewer 3: ```suggestion
First, be sure you have an Enterprise license installed. Staff members should have access to the shared developer license, but you can also request a trial license in-product. Browse to `/admin_console/about/license` on your local Mattermost instance to set up the license.
```

Second, enable performance monitoring. You can do this manually via `config.json` and setting `MetricsSettings.Enable` to `true`, or by exporting `MM_METRICSSETTINGS_ENABLE=true` into your shell before starting the server, or by enabling this manually via the System Console at `/admin_console/environment/performance_monitoring`:

![Performance monitoring configuration](/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana/performance-monitoring-config.png)

### What's next?

> These changes weren't just made in isolation -- in fact, this is all just infrastructure work for another recent project to improve Mattermost Performance Monitoring. Stay tuned for a follow-up blog post!

Reviewer 3: ```suggestion
These changes weren't just made in isolation - in fact, this is all just infrastructure work for another recent project to improve Mattermost performance monitoring. Stay tuned for a follow-up blog post!
```
